src/core/experimental/memutils.d

```diff
@@ -0,0 +1,332 @@
+    Dmemset_testStaticType!(double)(5);
+    Dmemset_testStaticType!(real)(5);
+    Dmemset_testDynamicArray!(ubyte)(5, 3);
+    static foreach (i; 1..10) {
+        Dmemset_testDynamicArray!(ubyte)(5, 2^^i);
+        Dmemset_testStaticArray!(ubyte, 2^^i)(5);
+    }
+    Dmemset_testDynamicArray!(ubyte)(5, 100);
+    Dmemset_testStaticArray!(ubyte, 100)(5);
+    Dmemset_testDynamicArray!(ubyte)(5, 500);
+    Dmemset_testStaticArray!(ubyte, 500)(5);
+    Dmemset_testDynamicArray!(ubyte)(5, 700);
+    Dmemset_testStaticArray!(ubyte, 700)(5);
+    Dmemset_testDynamicArray!(ubyte)(5, 3434);
+    Dmemset_testStaticArray!(ubyte, 3434)(5);
+    Dmemset_testDynamicArray!(ubyte)(5, 7128);
+    Dmemset_testStaticArray!(ubyte, 7128)(5);
+    Dmemset_testDynamicArray!(ubyte)(5, 13908);
+    Dmemset_testStaticArray!(ubyte, 13908)(5);
+    Dmemset_testDynamicArray!(ubyte)(5, 16343);
+    Dmemset_testStaticArray!(ubyte, 16343)(5);
+    Dmemset_testDynamicArray!(ubyte)(5, 27897);
+    Dmemset_testStaticArray!(ubyte, 27897)(5);
+    Dmemset_testDynamicArray!(ubyte)(5, 32344);
+    Dmemset_testStaticArray!(ubyte, 32344)(5);
+    Dmemset_testDynamicArray!(ubyte)(5, 46830);
+    Dmemset_testStaticArray!(ubyte, 46830)(5);
+    Dmemset_testDynamicArray!(ubyte)(5, 64349);
+    Dmemset_testStaticArray!(ubyte, 64349)(5);
+}
+
+// From a very good Chandler Carruth video on benchmarking: https://www.youtube.com/watch?v=nXaxk27zwlk
+void escape(void* p)
+{
+    version (LDC)
+    {
+        import ldc.llvmasm;
+        __asm("", "r,~{memory}", p);
+    }
+    version (GNU)
+    {
+        asm { "" : : "g" p : "memory"; }
+    }
+}
+
+void Dmemset_verifyArray(T)(int j, const ref T[] a, const ubyte v)
+{
+    const ubyte *p = cast(const ubyte *) a.ptr;
+    for (size_t i = 0; i < a.length * T.sizeof; i++)
+    {
+        assert(p[i] == v);
+    }
+}
+
+void Dmemset_verifyStaticType(T)(const ref T t, const ubyte v)
+{
+    const ubyte *p = cast(const ubyte *) &t;
+    for (size_t i = 0; i < T.sizeof; i++)
+    {
+        assert(p[i] == v);
+    }
+}
+
+void Dmemset_testDynamicArray(T)(const ubyte v, size_t n)
+{
+    T[] buf;
+    buf.length = n + 32;
+
+    enum alignments = 32;
+    size_t len = n;
+
+    foreach (i; 0..alignments)
+    {
+        auto d = buf[i..i+n];
+
+        escape(d.ptr);
+        Dmemset(d, v);
+        Dmemset_verifyArray(i, d, v);
+    }
+}
+
+void Dmemset_testStaticArray(T, size_t n)(const ubyte v)
+{
+    T[n + 32] buf;
+
+    enum alignments = 32;
+    size_t len = n;
+
+    foreach (i; 0..alignments)
+    {
+        auto d = buf[i..i+n];
+
+        escape(d.ptr);
+        Dmemset(d, v);
+        Dmemset_verifyArray(i, d, v);
+    }
+}
+
+void Dmemset_testStaticType(T)(const ubyte v)
+{
+    T t;
+    escape(&t);
+    Dmemset(t, v);
+    Dmemset_verifyStaticType(t, v);
+}
+
+version (GNU)
+{
+    void Dmemset(void *d, const uint val, size_t n)
+    {
+        Dmemset_naive(d, cast(const(ubyte))val, n);
+    }
+}
+else
+{
+    version (D_SIMD)
```

Comment:
We should also provide a version if SIMD isn't supported

Reply:
I guess it was not clear, but I don't know a better way to do it.
If SIMD isn't supported, only `Dmemset_naive` will be compiled _and_ called.
Check the `void Dmemset(T)(ref T dst, const ubyte val)` definition (currently at the bottom of the file).

Reply:
OK, but if that's the SIMD version I would do keep SIMD versioning in the `memset` function. No reason to duplicate it.

Reply:
In which of 2? I see 2 possibilities:
a) Remove it altogether from there and just have it in `Dmemset(T)`
b) Put it inside the function instead of around it.

With a), the problem is that `Dmemset(void *...` will be compiled. And if SIMD is not available, it won't compile.
With b), without SIMD an empty function will be compiled. If that's good, ok.

Reply:
One variant:

```d
version(D_SIMD) memsetSIMD() {

...
}

auto memset(...) {

   version (D_SIMD)  {
     version (GNU) { memsetNaive(); }  else {
         memsetSIMD();
      }
   }
}
```

Another one:

```d
version (D_SIMD)  {
    version (GNU) {
       enum memsetWithSIMD = false;
    } else {
       memsetWithSIMD = true;
   }
} else {
   enum memsetWithSIMD = false;
} 

version(memsetWithSIMD) memsetSIMD() {
...
}

auto memset(...) {
     version (memsetWithSIMD) { 
       memsetSIMD();
      }  else {
        memsetNaive();
      }
   }
}
```

Reply:
Thanks, I did something like what you proposed. There is a problem though with GDC. Eventually, GDC should probably use intrinsics as well. That's not necessary, as @jpf91 said seems to be more interested in completely pure D implementation (Johannes you can give feedback here).

But if this happens, GDC's intrinsics are x86 only. Meaning, then the code will look like:
```
version (GNU)
{
    useSIMD = false;
    version (x86_64)
    {
         version (D_SIMD) { useSIMD = true; }
    }
    version (useSIMD)
    {
         // A same implementation as the current SIMD one, but with GDC intrinsics.
    }
    else
    {
        // What is done now - i.e. forward to naive.
    }
}
```
Which pretty much means that the code will get ugly.

Reply:
I made an update in the forums. GDC (and LDC if the SIMD version is not used) reaches complete parity with libc memset() in my benchmarks (AMD Ryzen). If that holds true for different systems (unfortunately, I don't have Intel to test on) then we could even use the SIMD version only for DMD.